packages/xen-api/src/index.js

```diff
@@ -937,75 +951,104 @@ export class Xapi extends EventEmitter {
     })
   }
 
-  _watchEvents() {
-    const loop = () =>
-      this.status === CONNECTED &&
-      pTimeout
-        .call(
-          this._sessionCall('event.from', [
-            ['*'],
-            this._fromToken,
-            EVENT_TIMEOUT + 0.1, // Force float.
-          ]),
-          EVENT_TIMEOUT * 1.1e3 // 10% longer than the XenAPI timeout
+  // - prevent multiple watches
+  // - swallow errors
+  async _watchEventsWrapper() {
+    if (!this._watching) {
+      this._watching = true
+      await ignoreErrors.call(this._watchEvents())
+      this._watching = false
+    }
+  }
+
+  // TODO: cancelation
+  async _watchEvents() {
+    this._clearObjects()
+
+    // compute the initial token for the event loop
+    //
+    // we need to do this before the initial fetch to avoid losing events
+    let fromToken
+    try {
+      fromToken = await this._sessionCall('event.inject', [
+        'pool',
+        this._pool.$ref,
+      ])
+    } catch (error) {
+      if (isMethodUnknown(error)) {
+        return this._watchEventsLegacy()
+      }
+    }
+
+    const types = this._watchedTypes || this._types
+
+    // initial fetch
+    await Promise.all(
+      types.map(async type => {
+        // FIXME: use _transportCall to avoid auto-reconnection
+        const { events } = await this._sessionCall('event.from', [
+          [type],
+          '',
+          0, // do not wait if no records for this type
+        ])
+
+        // we can bypass _processEvents here because they are all *add* event
+        // and all objects are of the same type
+        events.forEach(event =>
+          this._addObject(type, event.ref, event.snapshot)
         )
-        .then(onSuccess, onFailure)
+      })
+    )
+    this._resolveObjectsFetched()
 
-    const onSuccess = ({ events, token, valid_ref_counts: { task } }) => {
-      this._fromToken = token
-      this._processEvents(events)
+    // event loop
+    const debounce = this._debounce
+    while (true) {
+      if (debounce != null) {
+        await pDelay(this._debounce)
+      }
+
+      let result
+      try {
+        result = await this._sessionCall(
+          'event.from',
+          [types, fromToken, EVENT_TIMEOUT],
+          EVENT_TIMEOUT * 1e1
+        )
+      } catch (error) {
+        if (error instanceof TimeoutError) {
+          continue
+        }
+        if (areEventsLost(error)) {
```

Comment:
When does this occur?

Reply:
I'm not sure, I would guess when we take too long to fetch the events and XenServer no longer remembers the state at `fromToken` time.